### PR TITLE
Add permutation unittest 

### PR DESCRIPTION
**Pull request by dingquanyu**
1. fixed some naming errors in config.py e.g. template_single_embedder should have keys: c_in and c_out but it was c_in and c_m in the previous version
2. fixed incompatibilities between embedders.py and config.py This is mainly caused by some of the keys in config.py are not in embedders.py 
3. moved some matrix slicing and processing steps to numpy in loss.py to solve the integer overflow issue caused by GPU memory usage
4. wrote a lddt_ca_multimer function in loss.py specifically for calculating lddt loss in the case of multimer training so that it can accommodate multimer input data structure  
5. overwrote the validation matrix in train_openfold.py for data structures in multimer training mode

### CODE REVIEW

File: openfold/data/data_pipeline.py [hunk 1]
```diff
@@ -784,6 +784,45 @@ def read_template(start, size):
 
         return all_hits
 
+    def _parse_template_hits(
```
Comment (christinaflo) [inline openfold/data/data_pipeline.py:787]: Should also be gone after rebase

File: openfold/utils/loss.py [hunk 2]
```diff
@@ -1033,13 +1054,16 @@ def between_residue_clash_loss(
...
         neighbour_mask = neighbour_mask & (asym_id[..., :, None] == asym_id[..., None, :])
 
-    neighbour_mask = neighbour_mask[..., None, None]
+    neighbour_mask = neighbour_mask.detach().to('cpu').numpy()[..., None, None]
 
     c_n_bonds = (
         neighbour_mask
-        * c_one_hot[..., None, None, :, None]
-        * n_one_hot[..., None, None, None, :]
+        * c_one_hot.detach().to('cpu').numpy()[..., None, None, :, None]
```
Comment (christinaflo) [inline openfold/utils/loss.py:1061]: When I add my changes and you rebase, just make sure to remove all the cpu stuff before submitting PR

File: openfold/data/data_pipeline.py (hunk 1 above)
Comment (dingquanyu) [reply openfold/data/data_pipeline.py:787]: Ah sorry this file is outdated. I will update data_pipeline.py with the version on multimer branch now and push it.

File: openfold/utils/loss.py (hunk 2 above)
Comment (dingquanyu) [reply openfold/utils/loss.py:1061]: Sure

File: openfold/utils/loss.py [hunk 3]
```diff
@@ -504,6 +503,29 @@ def lddt_ca(
...
+    return lddt(
+        all_atom_pred_pos,
+        all_atom_positions,
+        all_atom_mask,
+        cutoff=cutoff,
+        eps=eps,
+        per_residue=per_residue,
+    )
+
+def lddt_ca_multimer(
```
[the remark this answers is not in the document]
Comment (dingquanyu) [reply openfold/utils/loss.py:516]: If it uses the same way as lddt_ca : ```all_atom_mask[..., ca_pos : (ca_pos + 1)]```  then the dimension is not correct. I have to perform this ``` all_atom_mask[..., None] - all_atom_mask[..., None, :]``` so that ```all_atom_mask``` ends up with a shape of ```[batch_size,n_res,n_res] ```

File: openfold/utils/loss.py [hunk 4]
```diff
@@ -2021,7 +2052,8 @@ def __init__(self, config):
         super(AlphaFoldMultimerLoss, self).__init__(config)
         self.config = config
 
-    def multi_chain_perm_align(self,out, batch, labels, shuffle_times=2):
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (dingquanyu) [reply openfold/utils/loss.py:2055]: ok I see

File: train_openfold.py [hunk 5]
```diff
@@ -61,6 +61,63 @@ def __init__(self, config):
...
+        outputs, 
+        superimposition_metrics=False
+    ):
+        metrics = {}
+        
+        gt_coords = batch["all_atom_positions"][...,-1]
+        pred_coords = outputs["final_atom_positions"]
+        all_atom_mask = batch["all_atom_mask"][...,-1]
+        all_atom_mask = all_atom_mask.unsqueeze(-1).expand(*all_atom_mask.shape,3)
+        # In the case of multimer training, no need to introduce an empty dimension to all_atom_mask
```
[the remark this answers is not in the document]
Comment (dingquanyu) [reply train_openfold.py:75]: Ah sorry I wrongly copied pasted ```_compute_validation_metrics``` here again in line 64. I'll removed this chunk

File: train_openfold.py [hunk 6]
```diff
@@ -272,6 +329,63 @@ def __init__(self, config):
         self.cached_weights = None
         self.last_lr_step = -1
 
+    def _compute_validation_metrics(self,
```
Comment (dingquanyu) [inline train_openfold.py:332]: @christinaflo This ```_compute_validation_metrics``` is overwrites the parent class method so that it uses ```lddt_ca_multimer``` instead of ```lddt_ca```  and I turned ```all_atom_mask```, ```ca_pos``` and ```all_atom_mask_ca``` to None once they are not used anymore just to free up memory on GPU
